leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java

```diff
@@ -108,23 +119,109 @@ public T buildResponse(final Response coapResponse) {
         final Request coapRequest = coapRequestBuilder.getRequest();
 
         // Add CoAP request callback
-        coapRequest.addMessageObserver(new AsyncRequestObserver<T>(coapRequest, destination, responseCallback,
-                errorCallback) {
-            @Override
-            public T buildResponse(final Response coapResponse) {
-                // Build LwM2m response
-                final LwM2mResponseBuilder<T> lwm2mResponseBuilder = new LwM2mResponseBuilder<T>(coapRequest,
-                        coapResponse, client, model, observationRegistry);
-                request.accept(lwm2mResponseBuilder);
-                return lwm2mResponseBuilder.getResponse();
-            }
-        });
+        coapRequest.addMessageObserver(
+                new AsyncRequestObserver<T>(coapRequest, destination, responseCallback, errorCallback) {
+                    @Override
+                    public T buildResponse(final Response coapResponse) {
+                        // Build LwM2m response
+                        final LwM2mResponseBuilder<T> lwm2mResponseBuilder = new LwM2mResponseBuilder<T>(coapRequest,
+                                coapResponse, client, model, observationRegistry);
+                        request.accept(lwm2mResponseBuilder);
+                        return lwm2mResponseBuilder.getResponse();
+                    }
+                });
+
+        // Store pending request to cancel it on deregistration
+        addPendingRequest(destination.getRegistrationId(), coapRequest);
 
         // Send CoAP request asynchronously
         final Endpoint endpoint = getEndpointForClient(destination);
         endpoint.sendRequest(coapRequest);
     }
 
+    private String getFloorKey(String registrationId) {
+        if (registrationId == null)
```

Comment:
This check is not needed. If this method returns `null`, a NPE will be raised later (`concurrentSkipListMap.subMap()`).
The null-check can be moved to the `cancelPendingRequests()` method.
Same thing for `getCeilingKey()`.